async-queue/src/fifo.rs

```diff
@@ -0,0 +1,403 @@
+        // 1. Load the current offsets.
+        let current = unsafe {
+            let offsets = (*descriptor.offsets).load(Ordering::SeqCst);
+            Offsets::new(offsets, descriptor.len as u32)
+        };
+        let was_empty = current.is_empty();
+
+        // 2. If the queue is full, wait, then go to step 1.
+        if current.is_full() {
+            return Err(TrySendError::QueueFull);
+        }
+
+        // 3. Add 1 to the write offset and do an atomic compare-and-swap (CAS)
+        //    with the current offsets. If the CAS was not succesful, go to step 1.
+        let new = current.increment_write_offset();
+        let current = current.as_usize();
+        let prev = unsafe {
+            (*descriptor.offsets).compare_and_swap(current, new.as_usize(), Ordering::SeqCst)
+        };
+        if prev == current {
+            break (new, was_empty);
+        }
+    };
+
+    // 4. Write the data, then the `id`.
+    let slot = unsafe { &mut *descriptor.data.add(new.write_offset()) };
+    slot.copy_except_id(&val);
+    slot.set_id(val.get_id());
+
+    // 5. If the queue was empty in step 1, signal the reader to wake up.
+    Ok(was_empty)
+}
+
+pub(crate) fn try_recv_impl<T: WithAtomicId>(descriptor: &FifoDescriptor<T>) -> Result<(T, /*wake up writer:*/ bool), TryRecvError> {
+    // 1. Load the current offsets.
+    let current = unsafe {
+        let offsets = (*descriptor.offsets).load(Ordering::SeqCst);
+        Offsets::new(offsets, descriptor.len as u32)
+    };
+    let was_full = current.is_full();
+
+    // 2. If the queue is empty, wait, then go to step 1.
+    if current.is_empty() {
+        return Err(TryRecvError::QueueEmpty);
+    }
+
+    // 3. Add 1 to the read offset.
+    let new = current.increment_read_offset();
+
+    let slot = loop {
+        // 4. Read the `id` at the new read offset.
+        let slot = unsafe { &mut *descriptor.data.add(new.read_offset()) };
+        let id = slot.get_id();
+
+        // 5. If `id` is `0`, go to step 4 (spin). Spinning is OK because data is
+        //    expected to be written imminently.
+        if id != 0 {
+            break slot;
+        }
+    };
+
+    // 6. Read the data, then store `0` in the `id`.
+    let mut val = T::empty();
+    val.copy_except_id(slot);
+    val.set_id(slot.get_id());
+    slot.set_id(0);
+
+    // 7. Store the new read offset.
+    let after = unsafe {
+        fetch_adjust(
+            &*descriptor.offsets,
+            new.read as isize - current.read as isize,
+            Ordering::SeqCst,
+        )
+    };
+
+    // 8. If the queue was full in step 1, signal the writer to wake up.
+    //    ... or became full during read
+    let became_full = Offsets::new(after, descriptor.len as u32).is_full();
+    Ok((val, was_full || became_full))
+}
+
+lazy_static! {
+    pub(crate) static ref RECEIVER_TRACKER: ReceiverTracker = ReceiverTracker::new();
+}
+
+pub(crate) struct ReceiverTracker(Mutex<HashSet<usize>>);
```

Comment:
`lazy_static` definition above requires the type to be `Send`. Is this somehow safer?

I think the better alternative is to keep this as is and change the signature of `new_receiver`/`drop_receiver` to take a real pointer type instead of `usize`

Reply:
I ended up removing this whole thing